New feature - Sampler aggregation

markharwood opened the pull request:
Used to limit any nested aggregations' processing to a sample of the top-scoring documents.

Optionally, a “diversify” setting can limit the number of collected matches that share a common value such as an "author".

The original "DeferringBucketCollector" is now abstracted with the bulk of the original code in new subclass BestBucketsDeferringCollector and the new alternative policy for deferring is implemented in the BestDocsDeferringCollector subclass.

The diversifying logic is reliant on Lucene 5.1 which has changes to support this specialized form of result collection.

Closes #8108

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java:
```diff
@@ -0,0 +1,236 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class BestBucketsDeferringCollector extends DeferringBucketCollector {
```
Now that we have multiple implementations of DeferringBucketCollector, could we have a class-level Javadoc on the implementation to describe briefly what each one is trying to achieve?

colings86 commented on docs/reference/search/aggregations/bucket/sampler-aggregation.asciidoc:
```diff
@@ -0,0 +1,159 @@
+[[search-aggregations-bucket-sampler-aggregation]]
+=== Sampler Aggregation
+
+A filtering aggregation used to limit any nested aggregations' processing to a sample of the top-scoring documents.
```
Can we say sub-aggregation here so we don't overload 'nested aggregation' which could cause confusion?

markharwood replied:
agree

colings86 commented on docs/reference/search/aggregations/bucket/sampler-aggregation.asciidoc:
```diff
@@ -0,0 +1,159 @@
+use a choice of key (e.g. `author`) which is loaded from a regular `field` or derived dynamically by a `script`.
+For the given choice of diversity key you can control the maximum number of documents collected on a shard with this key.
+
+Note that the aggregation will throw an error if the choice of key field or script produces multiple values for a document.
```
Should this be an ASCIIDoc note section?

markharwood replied:
Agreed. I'll separate out the reference material and put the anecdotal stuff in a note.

colings86 commented on docs/reference/search/aggregations/bucket/sampler-aggregation.asciidoc:
```diff
@@ -0,0 +1,159 @@
+    }
+}
+--------------------------------------------------
+Note in the above example we use the default `max_docs_per_key` setting of 1 and combine author and genre fields to ensure 
```
Should this be an ASCIIDoc note section?

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/BestDocsDeferringCollector.java:
```diff
@@ -0,0 +1,301 @@
+     * been collected directly.
+     */
+    @Override
+    public Aggregator wrap(final Aggregator in) {
```
Is this the same as in the other implementation? Maybe we should move this to the parent class as the default implementation of the wrap(Aggregator) method?

markharwood replied:
Before creating the PR I had to diff it to remind myself if there was any differences and there was. I could create a base class for the wrapped Aggregator that holds all of what is in common but creating another class seemed like overkill. Could revisit that decision.

markharwood commented:
@colings86 rebased on latest master if you get a chance to review

colings86 commented on docs/reference/search/aggregations/bucket/sampler-aggregation.asciidoc:
```diff
@@ -0,0 +1,160 @@
+[[search-aggregations-bucket-sampler-aggregation]]
+=== Sampler Aggregation
```
Not sure if it already is, but can't see it if so: we should mark this feature as experimental in the docs

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/DiversityBuilder.java:
```diff
@@ -0,0 +1,66 @@
+    int maxDocsPerKey = SamplerParser.MAX_DOCS_PER_KEY_DEFAULT;
+    String executionHint = null;
+
+    protected DiversityBuilder() {
```
This should have a parameter for the name so that users can name their diversity aggregation

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/DiversityBuilder.java:
```diff
@@ -0,0 +1,66 @@
+        return builder;
+    }
+
+    public void exposeInternalXContent(XContentBuilder builder, Params params) throws IOException {
```
What are you trying to reuse from the ValuesSourceAggregationBuilder? If it's the script parameters then it would be better to use the ScriptParameterParser for this instead

markharwood replied:
It's the whole logic for determining where values will be obtained (field or script or script with params etc). I also want to benefit from any future additions to value source capabilities.

It feels ugly here because there's an existing assumption that ValueSource definitions are always bound to an aggregation. In this case we have an optional clause in an agg (the diversify logic) where the reason for accessing a ValueSource is purely to enhance the filtering logic of my existing sampler agg, not create new child buckets based on values.

colings86 replied:
I see. Can the sampler aggregation not contain this ValueSource logic and only use it if the diversity is specified. If the ValueSource isn't used I think it calls createUnmapped() in the Factory so that method could create a sampler agg with no diversity and the create() method could create the sampler agg with diversity?

markharwood replied:
Several possible solutions:
1) Existing solution (ab/re)uses the builder which is designed as a base class for aggs
2) Cut-and-paste solution copies the ValuesSourceAggregationBuilder logic into DiversityBuilder
3) Move common builder code to a new class reused by both ValuesSourceAggregationBuilder and DiversityBuilder (over complicated?)
4) Make SamplerAggregationBuilder extend ValuesSourceAggregationBuilder but this moves optional value source definition up to a top-level clause in sample agg but I think it makes more sense in the "diversify" container object. 

Option 2 might be least-worst?

colings86 replied:
Actually I kind of like option 3 as then we would have a ValuesSourceBuilder to go with the ValuesSourceParser. I don't think the parser side relies on being nested at the same level as the agg so the builder side shouldn't either. I also think that this way we wouldn't need to have the DiversityBuilder which would be nice since there is no DiversityParser on the other side (all the parsing is done by the SamplerParser) The ValuesSourceAggregationBuilder and SamplerBuilder can just use rather than extend the ValuesSourceBuilder. WDYT?

markharwood replied:
>can just use rather than extend the ValuesSourceBuilder. WDYT?

As a wrapper rather than a subclass they just end up reproducing all the methods on ValuesSourceBuilder with delegating methods? Would be a no-brainer if ValueSourceBuilder had some complex logic but with it being little more than a state-holder it seems less valuable. Any additions to ValueSource type settings would also require extra delegating calls and there'd be no automatic compile fails if none of these new features were surfaced in the wrapper classes (unless we make the wrapper classes implement a common interface?)
.

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/SamplerAggregationBuilder.java:
```diff
@@ -0,0 +1,80 @@
+            // in ValuesSourceAggregationBuilder here for the "diversify" part
+            // of the sampler agg's criteria.
+            builder.field("diversify");
+            diversityBuilder.exposeInternalXContent(builder, params);
```
See comment above

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/BestDocsDeferringCollector.java:
```diff
@@ -0,0 +1,239 @@
+        Arrays.sort(sd, new Comparator<ScoreDoc>() {
+            @Override
+            public int compare(ScoreDoc o1, ScoreDoc o2) {
+                return o1.doc - o2.doc;
```
Will this not sort in descending doc order? do we not want it in ascending doc order? or does it not matter?

colings86 commented on src/main/java/org/elasticsearch/search/aggregations/bucket/sampler/SamplerParser.java:
```diff
@@ -0,0 +1,104 @@
+
+        ValuesSourceConfig vsConfig = vsParser.config();
+        if (vsConfig.valid()) {
+            return new SamplerAggregator.DiversifiedFactory(aggregationName, shardSize, executionHint, vsConfig, maxDocsPerValue);
```
should we check here that the executionHint is null if the vsConfig is for a numericField, since executionHints are not available for numeric fields?

markharwood replied:
I added the code and tests then had a change of heart.
The docs talk about the fact that this is just a hint and that it
> "will ignore the choice of execution hint if it is not applicable" 

It seems wrong then to throw an exception when it doesn't apply to numeric fields.

markharwood replied:
@jpountz This PR has 3 modes for execution_hint when it comes to managing keys used for de-duping (required for diversity). It adds code and also complexity to user docs plus the debate above over if we should error when hint is irrelevant e.g. when field is numeric.
I saw marginal differences in performance between the various modes on my 1m tweet index. Do you think it's worth keeping execution_hint as an option?

colings86 commented on src/test/java/org/elasticsearch/search/aggregations/bucket/SamplerTests.java:
```diff
@@ -0,0 +1,192 @@
+        assertThat(sample.getDocCount(), greaterThan(0l));
+        Terms authors = sample.getAggregations().get("authors");
+        assertThat(authors.getBuckets().size(), greaterThan(0));
+    }
```
Probably a good idea to add tests for each execution mode

markharwood replied:
I have code in this test that selects a random choice of execution hint already

colings86 commented:
@markharwood Left some comments

markharwood commented:
@colings86 Thanks for the review. I added a couple of comments above on execution_hint test coverage and updated the code based on your other comments.

markharwood commented:
@jpountz @clintongormley This PR allows users to do analytics on a sample where you can also choose to diversify results on the basis of a particular field (e.g. analyse top X tweets but no more than Y tweets from a single Twitter account on each shard).

The question is what is the least-worst thing to do on each shard given the unmapped problem ie the choice of diversifying field doesn't exist on one of the indexes/shards being queried:
1) Throw an error
2) Return no results (because we can't guarantee diversification)
3) Return top results but without applying any of the diversity constraints

markharwood commented:
Took a decision with Colin on the 2 remaining questions:
1) execution_hint follows the precedent set in terms agg - never errors e.g. when used in relation to a numeric field in which case is ignored
2) Unmapped choices of diversifying field will return no results rather than a sample of undiversified results

markharwood commented:
Poke @colings86

colings86 commented:
LGTM

markharwood commented:
Pushed to master https://github.com/elastic/elasticsearch/commit/63db34f649d1de038c30d61f3c5e17e059b19b69
